### PR TITLE
Fix Request Charset Issues in Web Cmdlets

### LINKED ISSUE #7618: Wrong encode in Invoke-RestMethod -Body (problem with cyrilic symbols)
Steps to reproduce
------------------

```powershell
Invoke-RestMethod -Uri $URI -Method Put -ContentType 'application/json' -Body '{"jsonrpc":"2.0","method":"get.object","params":{"latname":"testing","cyrname":"проверка"}}'

```

Expected behavior
-----------------

http request json
```none
{"jsonrpc":"2.0","method":"get.object","params":{"latname":"testing","cyrname":"проверка"}}
```

Actual behavior
---------------

http request json
```none
{"jsonrpc":"2.0","method":"get.object","params":{"latname":"testing","cyrname":"????????"}}
```

Environment data
----------------
Name                           Value
----                           -----
PSVersion                      6.1.0-rc.1
PSEdition                      Core
GitCommitId                    6.1.0-rc.1
OS                             Microsoft Windows 10.0.17134
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

Description
---------------

It works good if -InFile Parameter is used

```powershell
Invoke-RestMethod -Uri $URI -Method Put -ContentType 'application/json' -InFile $FilePath 
```
File must be encoded in UTF-8 with no BOM

I think cmdlet encodes body to ASCII
In PS 5.1, 6.0 the same bug is.
Thanks

### PR DESCRIPTION
**Pull request by markekraus**
## PR Summary

Closes #7618

Instantiating a new `MediaTypeHeaderValue` object fails when the `-ContentType` parameter includes a charset such as `application/json; charset=utf-8`. This makes it impossible to set the content encoding on web requests. Moving to `Parse()` ensures we actually get a proper `MediaTypeHeaderValue` when the charset is present, thus allowing users to set their request encoding via proper `-ContentType` values.

## PR Context  

#7618
#8736

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.  
- **User-facing changes**
    - [X] Not Applicable
    - **OR**  
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [X] [Add `[feature]` to your commit messages if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by markekraus** [comment]
The codefactor error is nowhere near the code I touched. *shrugs

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 [hunk 1]
```diff
@@ -482,6 +482,18 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
...
     }
 
+    It "Invoke-WebRequest supports sending request as UTF-8." {
+        $uri = Get-WebListenerUrl -Test 'POST'
+        $command = "Invoke-WebRequest -Uri '$uri' -Body 'проверка' -ContentType 'application/json; charset=utf-8' -Method 'POST'"
+
+        $result = ExecuteWebCommand -command $command
+        ValidateResponse -response $result
+
+        $Result.Output.Encoding.BodyName | Should -Be 'utf-8'
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: Please use `-BeExactly` for strings.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 [hunk 2]
```diff
@@ -482,6 +482,18 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
...
+        $object = $Result.Output.Content | ConvertFrom-Json
+        $object.Data| Should -BeExactly 'проверка'
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]:
```suggestion
        $object.Data | Should -BeExactly 'проверка'
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 [hunk 3]
```diff
@@ -482,6 +482,18 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
         $Result.Output.Content | Should -Match '⡌⠁⠧⠑ ⠼⠁⠒  ⡍⠜⠇⠑⠹⠰⠎ ⡣⠕⠌'
     }
 
+    It "Invoke-WebRequest supports sending request as UTF-8." {
+        $uri = Get-WebListenerUrl -Test 'POST'
+        $command = "Invoke-WebRequest -Uri '$uri' -Body 'проверка' -ContentType 'application/json; charset=utf-8' -Method 'POST'"
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: We could add a protection comment that it is important to have non-ASCII 'проверка'.
For Invoke-WebRequest below too.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 4]
```diff
@@ -1674,7 +1674,7 @@ internal long SetRequestContent(HttpRequestMessage request, string content)
                 // would be used if Charset is not supplied in the Content-Type property.
                 try
                 {
-                    var mediaTypeHeaderValue = new MediaTypeHeaderValue(ContentType);
+                    var mediaTypeHeaderValue = MediaTypeHeaderValue.Parse(ContentType);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: We could use MediaTypeHeaderValue.TryParse()

Comment (markekraus) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: No need. we already have logic in place to handle the exceptions and perform the correct operations.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 (hunk 3 above)
Comment (markekraus) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: fixed

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 (hunk 2 above)
Comment (markekraus) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: fixed

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 (hunk 1 above)
Comment (markekraus) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: fixed

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs (hunk 4 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: We should _avoid_ exceptions if we can. If the exceptions come only from the Parse() method we should use TryParse().

Comment (markekraus) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: Except that we have different behavior depending on the type of exception. We would hand to build our own logic to figure that out. When the type of exception determines behavior it is perfectly acceptable to use them in a try/catch.

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: My comment is not blocking.

**Review by SteveL-MSFT — APPROVED** [review]

**Review by TravisEz13 — APPROVED** [review]

**Comment by iSazonov** [comment]
@markekraus Thanks for fixing the issue!

**Pull request merged**